controllers/pod_controller.go

```diff
@@ -0,0 +1,190 @@
+func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	filter, err := predicate.LabelSelectorPredicate(v1.LabelSelector{
+		MatchExpressions: []v1.LabelSelectorRequirement{
+			{
+				Key:      instanceLabel,
+				Operator: v1.LabelSelectorOpExists,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}, builder.WithPredicates(filter)).
+		Complete(r)
+}
+
+func (r *PodReconciler) strategicMergePatch(ctx context.Context, obj client.Object, patch map[string]any) error {
+	jp, err := json.Marshal(patch)
+	if err != nil {
+		return fmt.Errorf("failed to marshal patch: %w", err)
+	}
+	return r.Client.Patch(ctx, obj, client.RawPatch(types.StrategicMergePatchType, jp))
+}
+
+func recordSchedulingMetrics(l logr.Logger, instance string, pod *corev1.Pod, timedOut bool) error {
+	tr, err := getTrackedStates(pod)
+	if err != nil {
+		return fmt.Errorf("failed to get state tracking timestamps: %w", err)
+	}
+
+	createdTime, ok := tr[podstate.PodCreated]
+	if !ok {
+		return fmt.Errorf("pod created time not found")
+	}
+
+	metricLabels := prometheus.Labels{"namespace": pod.Namespace, "name": instance}
+
+	cm := podTimeCompleted.MustCurryWith(metricLabels)
+	if timedOut {
+		cm.WithLabelValues(metricTimedOutLabel).Observe(time.Since(createdTime).Seconds())
+	} else {
+		cm.WithLabelValues(metricCompletedLabel).Observe(time.Since(createdTime).Seconds())
+	}
+
+	acknowledgedTime, hasAcknowledgedTime := tr[podstate.PodAcknowledged]
+	scheduledTime, hasScheduledTime := tr[podstate.PodScheduled]
+	waitingTime, hasWaitingTime := tr[podstate.PodWaiting]
+
+	usm := podTimeUnscheduled.MustCurryWith(metricLabels)
+	if hasScheduledTime {
+		usm.WithLabelValues(metricCompletedLabel).Observe(scheduledTime.Sub(createdTime).Seconds())
+	} else if hasAcknowledgedTime {
+		usm.WithLabelValues(metricCompletedLabel).Observe(acknowledgedTime.Sub(createdTime).Seconds())
+	} else if hasWaitingTime {
+		usm.WithLabelValues(metricCompletedLabel).Observe(waitingTime.Sub(createdTime).Seconds())
+	} else if timedOut {
+		usm.WithLabelValues(metricTimedOutLabel).Observe(time.Since(createdTime).Seconds())
+	}
+
+	uam := podTimeUntilAcknowledged.MustCurryWith(metricLabels)
+	if hasAcknowledgedTime {
+		uam.WithLabelValues(metricCompletedLabel).Observe(acknowledgedTime.Sub(createdTime).Seconds())
+	} else if hasWaitingTime {
+		uam.WithLabelValues(metricCompletedLabel).Observe(waitingTime.Sub(createdTime).Seconds())
+	} else if timedOut {
+		uam.WithLabelValues(metricTimedOutLabel).Observe(time.Since(createdTime).Seconds())
+	}
+
+	ptuw := podTimeUntilWaiting.MustCurryWith(metricLabels)
+	if hasWaitingTime {
+		ptuw.WithLabelValues(metricCompletedLabel).Observe(waitingTime.Sub(createdTime).Seconds())
+	} else if timedOut {
+		ptuw.WithLabelValues(metricTimedOutLabel).Observe(time.Since(createdTime).Seconds())
+	} else {
+		return fmt.Errorf("pod has no waiting time")
+	}
```

Comment:
Nitpick: These might be a bit more readable if combined into single switch statement instead of spreading it out into three if elseif blocks

Reply:
it kinda melts my brain a bit atm how this switch would look 😅 i'll try reformatting with some additional tests when i have time this afternoon